hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/ConsensusSubmitMessageHandler.java

```diff
@@ -68,22 +89,162 @@ public void preHandle(@NonNull final PreHandleContext context, @NonNull Readable
+            }
+
+            /* Validate if the transaction is submitting initial chunk,payer in initial transaction Id should be same as payer of the transaction */
+            if (1 == chunkInfo.getNumber()
+                    && !chunkInfo.getInitialTransactionID().equals(txnId)) {
+                throw new HandleStatusException(INVALID_CHUNK_TRANSACTION_ID);
+            }
+        }
+    }
+
+    /**
+     * Updates the running hash and sequence number of the topic.
+     * @param txn the {@link TransactionBody} of the active transaction
+     * @param topic the topic to which the message is being submitted
+     * @param consensusNow the consensus time of the active transaction
+     * @return the updated topic
+     * @throws IOException if there is an error while updating the running hash
+     */
+    public Topic updateRunningHashAndSequenceNumber(final TransactionBody txn, final Topic topic, Instant consensusNow)
+            throws IOException {
+        final var payer = txn.getTransactionID().getAccountID();
+        final var topicId = txn.getConsensusSubmitMessage().getTopicID();
+        final var message = txn.getConsensusSubmitMessage().getMessage().toByteArray();
+
+        // This line will be uncommented once there is PBJ fix to make copyBuilder() public
+        // final var topicBuilder = topic.copyBuilder();
+        final var topicBuilder = new Topic.Builder()
+                .topicNumber(topic.topicNumber())
+                .adminKey(topic.adminKey())
+                .submitKey(topic.submitKey())
+                .autoRenewAccountNumber(topic.autoRenewAccountNumber())
+                .autoRenewPeriod(topic.autoRenewPeriod())
+                .expiry(topic.expiry())
+                .deleted(topic.deleted())
+                .memo(topic.memo());
+
+        if (null == consensusNow) {
+            consensusNow = Instant.ofEpochSecond(0);
+        }
+
+        var sequenceNumber = topic.sequenceNumber();
+        var runningHash = topic.runningHash();
+
+        final var boas = new ByteArrayOutputStream();
+        try (final var out = new ObjectOutputStream(boas)) {
+            out.writeObject(unwrapPbj(runningHash));
+            out.writeLong(RUNNING_HASH_VERSION);
+            out.writeLong(payer.getShardNum());
+            out.writeLong(payer.getRealmNum());
+            out.writeLong(payer.getAccountNum());
+            out.writeLong(topicId.getShardNum());
+            out.writeLong(topicId.getRealmNum());
+            out.writeLong(topicId.getTopicNum());
+            out.writeLong(consensusNow.getEpochSecond());
+            out.writeInt(consensusNow.getNano());
+
+            /* Update the sequence number */
+            topicBuilder.sequenceNumber(++sequenceNumber);
+
+            out.writeLong(sequenceNumber);
+            out.writeObject(noThrowSha384HashOf(message));
+            out.flush();
+            runningHash = Bytes.wrap(noThrowSha384HashOf(boas.toByteArray()));
```

Comment:
This logic could be tested a lot more easily and a lot more thoroughly if it were extracted to a serializer object